### PR TITLE
SNP-660: flutter secure storage

### PR DESCRIPTION
**Pull request by internetova**
## Changes

- Added `flutter_secure_storage` dependency
- Added AuthTokensStorage
- Added Failure & ServerFailure
- Added AuthTokenInterceptor, RefreshTokenInterceptor, RetryInterceptor, ServerErrorInterceptor
- Added RefreshTokenRepository template
- Added json_annotation
- Added how-to-use section in README.md for  FlutterSecureStorage

## Checklist for self-check
- [x] Commits and PRs have been filed according to [the rules on the project](https://github.com/surfstudio/surf-flutter-app-template#workflow-in-a-repository).
- [x] The author is marked as an assigne and assigned mandatory reviewers.
- [ ] Required labels marked
- [ ] Specified related tasks and/or related PRs.
- [ ] Specified Changes.
- [ ] Attached videos/screenshots demonstrating the fix/feature.
- [ ] All unspecified fields in the PR description deleted.
- [ ] New code covered by tests.

## Checklist for reviewers
- [ ] CI passed successfully _(with a green check mark)_.
- [ ] PR is atomic, by volume no more than 400 (+-) corrected lines (not including codogen).

Design:
- [ ] System design corresponds to the agreements on structure and architecture on the project.
- [ ] The code is decomposed into necessary and sufficient components.

Functionality:
- [ ] The code solves the problem.
- [ ] Any changes to the user interface are reasonable and look good.

Complexity:
- [ ] The code is clear, easy to read, functions are small, no more than 50 lines.
- [ ] The logic is not overcomplicated, there is no overengineering (no code sections that may be needed in the future, but no one knows about it).

Tests:
- [ ] Updated or added tests for mandatory components.
- [ ] The tests are correct, helpful, and well designed/developed.

Naming:
- [ ] The naming of variables, methods, classes and other components is understandable.

Comments:
- [ ] The comments are understandable and helpful.

Documentation:
- [ ] All labels are correct
- [ ] Technical documentation updated (after approval, updates last reviewer).

### CODE REVIEW

File: lib/features/common/service/repository/refresh_token/refresh_token_impl.dart [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      final failure = DioErrorMapper.mapToFailure(e);
+
+      if (failure is RefreshTokenFailure) {
+        // TODO(anyone): log write e
+        // TODO(anyone): handleRefreshTokenFailure
+      }
+    } on Exception catch (e) {
+      // TODO(anyone): log write e
+      // TODO(anyone): handleRefreshTokenFailure
+    }
```
Comment (maxim-sysoev) [inline lib/features/common/service/repository/refresh_token/refresh_token_impl.dart:46]: I guess we can face with token refreshing on parallel requests executing (two refresh token requsts will be executin parallel too, and one of them will finish with error). I think we should add completer for executing one refresh token request, and after it done - return value.

File: lib/features/common/service/mappers/dio_error_mapper.dart [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    if (e is Failure) {
+      return e;
+    }
+
+    return Failure(
+      parentException: error,
+      stackTrace: st,
+    );
+  }
+}
```
Comment (maxim-sysoev) [inline lib/features/common/service/mappers/dio_error_mapper.dart:18]: i guess there is must be mapping to concrete types of failure

File: lib/features/common/service/interceptors/refresh_token_interceptor.dart [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          requestEncoder: options.requestEncoder,
+          responseDecoder: options.responseDecoder,
+          listFormat: options.listFormat,
+        ),
+        data: options.data,
+        queryParameters: options.queryParameters,
+        cancelToken: options.cancelToken,
+        onSendProgress: options.onSendProgress,
+        onReceiveProgress: options.onReceiveProgress,
+      );
```
Comment (maxim-sysoev) [inline lib/features/common/service/interceptors/refresh_token_interceptor.dart:58]: is no way to do it like in https://github.com/surfstudio/surf-flutter-app-template/pull/54/files#diff-7542378919d78a18ff3e7f9d1560acf17081027f858df1b36e6c99a71ecf1b56R40

`final result = await _dio.fetch<void>(options);`

File: lib/features/app/di/app_scope.dart [hunk 4]
```diff
@@ -22,6 +30,12 @@ class AppScope implements IAppScope {
   @override
   late VoidCallback applicationRebuilder;
 
+  @override
+  final FlutterSecureStorage secureStorage = const FlutterSecureStorage();
+
+  @override
+  late final IAuthTokensStorage authTokensStorage = _createAuthTokensStorage();
```
Comment (maxim-sysoev) [inline lib/features/app/di/app_scope.dart:37]: can we simply set `= AuthTokensStorageSecure(secureStorage)` without using method?

File: lib/features/common/service/interceptors/refresh_token_interceptor.dart (hunk 3 above)
Comment (internetova) [reply lib/features/common/service/interceptors/refresh_token_interceptor.dart:58]: @maxim-sysoev dio

Comment (maxim-sysoev) [reply lib/features/common/service/interceptors/refresh_token_interceptor.dart:58]: and they have different result?

Comment (internetova) [reply lib/features/common/service/interceptors/refresh_token_interceptor.dart]: replaced dio.request with dio.fetch ✅

**Comment by maxim-sysoev** [comment]
i think we needs to test it on real backend, without this we cannot be sure what it's works

**Comment by internetova** [comment]
https://github.com/surfstudio/surf-flutter-app-template/pull/62

**Pull request closed without merging**